tinybc.1
.TH tinybc 1 "January 30, 2013" "" "Tiny BASIC for Curses"

.SH NAME
tinybc \- Tiny BASIC for Curses

..SH SYNOPSIS
.B tinybc
.RI [OPTIONS]  [FILE]
.br

.SH DESCRIPTION
\fBtinybc\fP is a BASIC interpreter for the curses character screen handling library, which fully corresponds to the Tiny BASIC specification.

The tinybc engine is thread-safe and can be embedded into other code. Embedding may not make sense because of the capabilities of the interpreter, but it makes sense because the code is easily extendable.

It's just an idea i got and quickly implemented, BASIC is not a programming language which i use. I made it for fun and i hope that it would be used for fun, such as a minimalist challenge.

.SH Curses
Curses is a library for using computer interactively in a character mode. It corresponds to the POSIX standard, and is the most cross-platform solution for that purpose. Character mode is the easiest to code, so it is the easiest to feel the computer that way. Terminals of the modern computers provide almost the same "graphics" as some earliest microcomputers, and there are many games with such graphics.

.SH Tiny BASIC
Tiny BASIC is the specification of the most minimal BASIC. Actually there is not much to choose from, because the next smallest specification is the Small BASIC standard, and this does not provide a small language at all. The Tiny BASIC specification was written by Dennis Allison and was published in the People's Computer Company newsletter in December 1975. The specification was necessary because of the need to have a small interpreter which can fit into 2 KB of memory. Tiny BASIC was one of the first programming languages for the microcomputers.

This interpreter is minimal not because it was difficult to add more functionality, adding functionality is surprisingly simple. This interpreter is minimal because it is meant to implement Tiny BASIC. But i think that minimalism enables creatity.

This interpreter is mostly based on the Tom Pittman's Tiny BASIC, except the USR function and using full expressions for input. The Tom Pittman's Tiny BASIC is implemented so completely that the programs written in it run without modifications. The computed GOTO and GOSUB are implemented too, but their expressions would be renumbered when they only consist of a single number. Two additional features are taken from the Palo Alto Tiny BASIC, one big array named @, and using relation operators in expressions. This enables to run programs written in Palo Alto Tiny BASIC with some modifications.

.SH OPTIONS
.IP \fB\-i\fP
Starts the interpreter interactively and loads the file.
.IP \fB\-number\fP
Renumbers the program starting from number.

.SH INSTALL
To compile tinybc in Linux, type make in the terminal. To install tinybc, as root type make install, and to remove the installation, type make remove. The installation also copies the examples to the directory /usr/share/doc/tinybc . To compile tinybc in Windows and DOS, type one of the following depending on your system:

mingw32-make tinybc.exe

make tinybc.exe

make dos

The binaries for Linux and Windows are provided, but because of viruses, etc, it is safer to compile. The program depends only on some version of the curses library. For compiling in Linux, the ncurses development package has to be installed. In Windows, the compiler such as MinGW, and in DOS DJGPP has to be installed. When you install the compiler, add its bin directory to the path. The PDCurses library for Windows is included, in DOS the library in \\pdcurses\\dos has to be compiled. In DOS, tinybc has only been tested in DOSBox, there the files were copied to the directory \\tinybc08 .

The path of the temporary directory should be set to the environment variable TINYBC_TMP , otherwise temporary files shall be stored in the current directory. The path must end with slash or backslash.

.SH LANGUAGE DESCRIPTION
The features such as an array were added in order to make tinybc as advanced as Tiny BASIC, not more advanced. Because in spite that the Tiny BASIC specification provides no arrays, most of the Tiny BASIC interpreters at the time (1976) enabled some memory access, thus providing a data storage similar to array. Such memory access also enabled to read program statements, which could been used to read pre-defined data. In tinybc, assigning strings and lists of expressions to array is provided to compensate the lack of DATA statement in Tiny BASIC.

These interpreters also had an access to input/output port, and they could also send escape sequences to the display. The statements OUT, INKEY, and the other curses statements provide that in tinybc. It was possible to run system functions, any output to paper tape and read any input from paper tape. The input/output redirection, copying and pasting to the terminal and the SYSTEM statement provide that functionality.

Because of the lack of the if and while blocks, Tiny BASIC does not correspond to the modern programming paradigm. Structured programming is nevertheless possible by marking all destinations of jumps with REM statements, such as REM SUBROUTINE DrawPixel (X, Y), REM DO, REM ELSE and REM END IF, and by using empty REM statements for making the code clearer. The original Tiny BASIC had no FOR...NEXT loop, which enables to write code faster, but does not significantly increase the speed of programming.

It is difficult to make the simplest language which is usable for some purposes and which still corresponds to some standard. Adding even a single seemingly simple feature can be enough to make the language not simple or not standard. Making a simple language is a matter of such delicate balance.

.SH Program statements
In statements, all lines must begin with a number. It is not always necessary though that the numbers are different or in a growing order. The line numbers are required because Tiny BASIC is a traditional BASIC. The lines can also be renumbered. See the examples (the files with the bas extension). The LET and THEN keywords can be omitted, the PRINT keyword can be abbreviated as PR.

.IP "\fBCLS\fP"
Clears the screen
.IP "\fBCOLOR expr1, expr2\fP"
Changes the foreground color to expr1 and background color to expr2
.IP "\fBEND\fP"
Ends the execution of the program, can be anywhere in the program
.IP "\fBGOSUB expr\fP"
Executes a subroutine which starts with the number which is the value of the expression, subroutine ends with RETURN
.IP "\fBGOTO expr\fP"
Executes the line with the number which is the value of the expression
.IP "\fBIF expr THEN statement\fP"
Executes statement if the value of the expression is not zero
.IP "\fBINCHAR var1, var2, var3\fP"
Writes the code of the character at the current location into var1, var2 and var3 are for color codes and can be omitted
.IP "\fBINPUT var, ... var\fP"
Inputs variables
.IP "\fBINKEY var\fP"
Inputs a key code, -1 if no keys pressed
.IP "\fBLET var = expr\fP"
Assigns an expression to variable
.IP "\fBLET @(expr) = string|expr, ... string|expr\fP"
Assigns the expressions and characters from strings to consecutive array elements
.IP "\fBLOCATE expr1, expr2\fP"
Moves cursor to line expr1 and column expr2, the next PRINT prints to that location
.IP "\fBNAP expr\fP"
Sleeps the expr of milliseconds, a NAP statement should be in the main loop to give time for the operating system
.IP "\fBOUT expr\fP"
Prints a character, the code of which is the expression
.IP "\fBPRINT string|expr, .... string|expr\fP"
Prints the strings and expressions, ; allowed
.IP "\fBREM comment\fP"
Any text can be written after REM as a comment
.IP "\fBRETURN\fP"
Ends the subroutine block and executes the line after the calling GOSUB statement
.IP "\fBSYSTEM expr1, expr2\fP"
Filters the text in array starting at expr2 through the external command starting at expr1, both end with 0

.SH Interactive mode
Interactive mode enables to both edit and run the program. Interactive mode is a part of the Tiny BASIC specification, so it has to be implemented in every language which is claimed to be Tiny BASIC. The HELP statement is added. LOAD and SAVE are also additional statements which are not in the Tiny BASIC specification. In the early computers they used teletypes as terminals, so there was no difference whether the input or ouput was done with a typewriter, or with a punched tape.

As it is the tradition in BASIC, the first character can be written instead of the full keyword.

If you write a line which starts with a number, then this is considered to be a program line and it will be added to to program to the appropriate place. If the number is the number of an existing statement, then this statement will be replaced with the statement which you entered. If the line consists only of a number with no additional text, then the statement with that number shall be deleted.

.IP "\fBCLEAR\fP"
Deletes the program
.IP "\fBLIST\fP"
Lists the whole program
.IP "\fBLIST line\fP"
Lists the line
.IP "\fBLIST line1-line2\fP"
Lists program from line1 to line2
.IP "\fBLOAD filename\fP"
Loads program from file
.IP "\fBQUIT\fP"
Exits
.IP "\fBRUN\fP"
Runs the program
.IP "\fBSAVE [filename]\fP"
Saves the program
.IP "\fBHELP\fP"
Prints help

.SH Variables

releases.txt
    * 0.8.6
    * 30 Jan 2013
Release Notes: This release makes the output faster and adds reading colors and 3D rendering example.

    * 0.8.4
    * 10 Jan 2013 21:22
Release Notes: This release adds support for DOS.

    * 0.8.2
    * 07 Jan 2013 23:37
Release Notes: This release adds utilities to read and write files in Windows.

    * 0.8.0
    * 27 Dec 2012 00:54
Release Notes: This release adds the statement SYSTEM for filtering through external commands.

    * 0.6.8
    * 05 Dec 2012 23:41
Release Notes: This release fixes the bugs in running Windows files and restoring terminal settings.

    * 0.6.6
    * 02 Dec 2012 01:31
Release Notes: This release includes the manual for beginners and fixes the colors and error messages.

    * 0.6.4
    * 01 Nov 2011 20:59
Release Notes: This release adds OUT and LOAD statements and assignment of strings to arrays.

    * 0.6.2
    * 29 Oct 2011 02:54
Release Notes: This release adds assignment of expression list to array.

    * 0.6.0
    * 21 Oct 2011 20:53
Release Notes: This release adds one big array named @, as in Palo Alto Tiny BASIC.

    * 0.5.0
    * 16 Oct 2011 01:05
Release Notes: This is the first release.



tinybc.c
/*
 * Tiny BASIC for Curses
 *
 * Copyright (c) 2011, Tarvo Korrovits
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA  02110-1301  USA
 *
 */

#include "tokenize.h"
#include "tinybc.h"

static long int expr(struct cttype *ct);
static void stmt(struct cttype *ct);

static long int factor(struct cttype *ct)
{
	long int minus, nind, n;

	if (t_ended(ct->t)) return 0;
	PDEBUG(lg, "factor: token %d\n", t_type(ct->t));
	if (t_type(ct->t) == TOKEN_MINUS) {
		t_take(ct->t, TOKEN_MINUS);
		minus = -1;
		PDEBUG(lg, "factor: minus\n");
	} else
		minus = 1;
	switch (t_type(ct->t)) {
	case TOKEN_RND:
		t_take(ct->t, TOKEN_RND);
		t_take(ct->t, TOKEN_LPAREN);
		n = minus * rand() % expr(ct);
		PDEBUG(lg, "factor: random number %d\n", n);
		t_take(ct->t, TOKEN_RPAREN);
		break;
	case TOKEN_SIZE:
		t_take(ct->t, TOKEN_SIZE);
		n = minus * t_size(ct->t);
		PDEBUG(lg, "factor: size %d\n", n);
		break;
	case TOKEN_NUMBER:
		n = minus * t_number(ct->t);
		PDEBUG(lg, "factor: number %d\n", n);
		t_take(ct->t, TOKEN_NUMBER);
		break;
	case TOKEN_LPAREN:
		t_take(ct->t, TOKEN_LPAREN);
		n = minus * expr(ct);
		PDEBUG(lg, "factor: expression %d\n", n);
		t_take(ct->t, TOKEN_RPAREN);
		break;
	case TOKEN_VARIABLE:
		n = minus * tinybc_get(ct, t_variable(ct->t));
		PDEBUG(lg, "factor: obtaining value %d from "
			"variable number %d\n",
			tinybc_get(ct, t_variable(ct->t)),
			t_variable(ct->t));
		t_take(ct->t, TOKEN_VARIABLE);
		break;
	case TOKEN_AT:
		t_take(ct->t, TOKEN_AT);
		t_take(ct->t, TOKEN_LPAREN);
		nind = expr(ct);
		n = minus * tinybc_get(ct, MAX_VARIND + nind);
		PDEBUG(lg, "factor: obtaining value %d from "
			"array element number %d\n",
			tinybc_get(ct, MAX_VARIND + nind), nind);
		t_take(ct->t, TOKEN_RPAREN);
		break;
	default:
		PDEBUG(lg, "factor: wrong token\n");
		t_end(ct->t);
		return 0;
	}
	return n;
}

static long int term(struct cttype *ct)
{
	long int n1, n2, op;

	if (t_ended(ct->t)) return 0;
	n1 = factor(ct);
	op = t_type(ct->t);
	PDEBUG(lg, "term: token %d\n", op);
	/* Does all multiplications and divisions */
	while (op == TOKEN_ASTR || op == TOKEN_SLASH) {
		t_next(ct->t);
		n2 = factor(ct);
		PDEBUG(lg, "term: %d %d %d\n", n1, op, n2);
		switch (op) {
		case TOKEN_ASTR:
			n1 = n1 * n2;
			break;
		case TOKEN_SLASH:
			n1 = n1 / n2;
			break;
		}
		op = t_type(ct->t);
	}
	PDEBUG(lg, "term: %d\n", n1);
	return n1;
}

static long int component(struct cttype *ct)
{
	long int n1, n2, op;
	
	if (t_ended(ct->t)) return 0;
	n1 = term(ct);
	op = t_type(ct->t);
	PDEBUG(lg, "component: token %d\n", op);
	while (op == TOKEN_PLUS || op == TOKEN_MINUS) {
		t_next(ct->t);
		n2 = term(ct);
		PDEBUG(lg, "component: %d %d %d\n", n1, op, n2);
		switch(op) {
		case TOKEN_PLUS:
			n1 = n1 + n2;
			break;
		case TOKEN_MINUS:
			n1 = n1 - n2;
			break;
		}
		op = t_type(ct->t);
	}
	PDEBUG(lg, "component: %d\n", n1);
	return n1;
}

static long int expr(struct cttype *ct)
{
	long int n1, n2, op;

	if (t_ended(ct->t)) return 0;
	n1 = component(ct);
	op = t_type(ct->t);
	PDEBUG(lg, "expr: token %d\n", op);
	while (op == TOKEN_LT || op == TOKEN_GT ||
		op == TOKEN_EQ || op == TOKEN_NOTEQ ||
		op == TOKEN_LE || op == TOKEN_GE) {
		t_next(ct->t);
		n2 = component(ct);
		PDEBUG(lg, "expr: %d %d %d\n", n1, op, n2);
		switch(op) {
		case TOKEN_LT:
			n1 = n1 < n2;
			break;
		case TOKEN_GT:
			n1 = n1 > n2;
			break;
		case TOKEN_EQ:
			n1 = n1 == n2;
			break;
		case TOKEN_NOTEQ:
			n1 = n1 != n2;
			break;
		case TOKEN_LE:
			n1 = n1 <= n2;
			break;
		case TOKEN_GE:
			n1 = n1 >= n2;
			break;
		}
		op = t_type(ct->t);
	}
	PDEBUG(lg, "expr: %d\n", n1);
	return n1;
}

void interactive(int mode)
{
	if (mode) {
		curs_set(0);
		cbreak();
		noecho();
		nodelay(stdscr, TRUE);
		keypad(stdscr, TRUE);
		scrollok(stdscr, FALSE);
	} else {
		curs_set(1);
		nocbreak();
		echo();
		nodelay(stdscr, FALSE);
		keypad(stdscr, FALSE);
		scrollok(stdscr, TRUE);
	}
}

#if 0
/*
 * Example of a new statement, assigns n2 to n1.
 * To add a new statement, do the following:
 * 1. Add the new token type to the enum in tokenize.h,
 * 2. Add the keyword to keywords array in tokenize.c,
 * 3. Add new statement to the statement function below,
 * 4. Include a new header file here if necessary,
 * 5. Write the new function below this example
 */
static void assign_stmt(struct cttype *ct)
{
	long int n1ind, n2;

	t_take(ct->t, TOKEN_ASSIGN);
	PDEBUG(lg, "assign:\n");
	if (t_type(ct->t) == TOKEN_VARIABLE) {
		n1ind = t_variable(ct->t);
		t_take(ct->t, TOKEN_VARIABLE);
	} else
		t_take(ct->t, TOKEN_VARIABLE);
	t_take(ct->t, TOKEN_COMMA);
	if (t_expr_type(ct->t)) {
		n2 = expr(ct);
		/* Now we have both parameters, do the operation */
		tinybc_set(ct, n1ind, n2);
	} else
		t_take(ct->t, TOKEN_VARIABLE);
	t_take(ct->t, TOKEN_LF);
}
#endif

static void at_stmt(struct cttype *ct)
{
	char string[MAX_STRLEN], *p;
	long int n1ind;

	t_take(ct->t, TOKEN_AT);
	PDEBUG(lg, "at:\n");
	t_take(ct->t, TOKEN_LPAREN);
	n1ind = expr(ct);
	t_take(ct->t, TOKEN_RPAREN);
	t_take(ct->t, TOKEN_EQ);
	for (; ;)
		if (t_type(ct->t) == TOKEN_STRING) {
			t_string(ct->t, string, MAX_STRLEN);
			for (p = string; *p; p++)
				tinybc_set(ct, MAX_VARIND + n1ind++, *p);
			t_take(ct->t, TOKEN_STRING);
		} else if (t_expr_type(ct->t))
			tinybc_set(ct, MAX_VARIND + n1ind++, expr(ct));
		else if (t_type(ct->t) == TOKEN_COMMA)
			t_take(ct->t, TOKEN_COMMA);
		else
			break;
	t_take(ct->t, TOKEN_LF);
}

static void cls_stmt(struct cttype *ct)
{
	t_take(ct->t, TOKEN_CLS);
	PDEBUG(lg, "cls:\n");
	bkgd(COLOR_PAIR(ct->currentcolor[0]));
	clear();
	t_take(ct->t, TOKEN_LF);
}

static void color_stmt(struct cttype *ct)
{
	long int n1, n2;

	t_take(ct->t, TOKEN_COLOR);
	PDEBUG(lg, "color:\n");
	if (t_expr_type(ct->t)) {
		n1 = expr(ct);
#ifdef BGR
		if (n1 == 1)
			n1 = 4;
		else if (n1 == 3)
			n1 = 6;
		else if (n1 == 4)
			n1 = 1;
		else if (n1 == 6)
			n1 = 3;
#endif
	} else
		t_take(ct->t, TOKEN_VARIABLE);
	t_take(ct->t, TOKEN_COMMA);
	if (t_expr_type(ct->t)) {
		n2 = expr(ct);
#ifdef BGR
		if (n2 == 1)
			n2 = 4;
		else if (n2 == 3)
			n2 = 6;
		else if (n2 == 4)
			n2 = 1;
		else if (n2 == 6)
			n2 = 3;
#endif
		/* Only COLOR_PAIRS color pairs are allowed */
		if (ct->colorpairs[n1][n2] && 
			n1 < MAX_COLOR && n2 < MAX_COLOR) {
			ct->currentcolor[0] = ct->colorpairs[n1][n2];
			ct->currentcolor[1] = n1;
			ct->currentcolor[2] = n2;
		} else if (ct->lastpair < COLOR_PAIRS - 1 &&
			n1 < MAX_COLOR && n2 < MAX_COLOR) {
			ct->lastpair++;
			ct->colorpairs[n1][n2] = ct->lastpair;
			ct->currentcolor[0] = ct->lastpair;
			ct->currentcolor[1] = n1;
			ct->currentcolor[2] = n2;
			init_pair(ct->currentcolor[0], n1, n2);
		}
		attrset(COLOR_PAIR(ct->currentcolor[0]));
	} else
		t_take(ct->t, TOKEN_VARIABLE);
	t_take(ct->t, TOKEN_LF);
}

static void end_stmt(struct cttype *ct)
{
	t_take(ct->t, TOKEN_END);
	PDEBUG(lg, "end:t\n");
	t_end(ct->t);
	t_take(ct->t, TOKEN_LF);
}

static void gosub_stmt(struct cttype *ct)
{
	long int linenum;

	t_take(ct->t, TOKEN_GOSUB);
	PDEBUG(lg, "gosub:\n");
	linenum = expr(ct);
	if (ct->gosub_stack_ptr < MAX_GOSUB_DEPTH) {
		ct->gosub_stack[ct->gosub_stack_ptr++] = t_number(ct->t);
		t_jump(ct->t, linenum);
	} else {
		printw("Gosub stack exhausted, line %d\n",
			t_currline(ct->t));
		t_end(ct->t);
	}
}

static void goto_stmt(struct cttype *ct)
{
	long int linenum;

	t_take(ct->t, TOKEN_GOTO);
	PDEBUG(lg, "goto:\n");
	linenum = expr(ct);
	t_jump(ct->t, linenum);
}

static void if_stmt(struct cttype *ct)
{
	long int n;
	
	t_take(ct->t, TOKEN_IF);
	n = expr(ct);
	PDEBUG(lg, "if: expression %d\n", n);
	/* THEN not compulsory */
	if (t_type(ct->t) == TOKEN_THEN) t_take(ct->t, TOKEN_THEN);
	if (n) {
		if (t_type(ct->t) == TOKEN_LET) t_take(ct->t, TOKEN_LET);
		stmt(ct);
	} else
		t_next_line(ct->t);
}

static void inchar_stmt(struct cttype *ct)
{
	long int n1ind, n2ind, n3ind, ch, co, i, j, k;

	t_take(ct->t, TOKEN_INCHAR);
	PDEBUG(lg, "inchar:\n");
	if (t_type(ct->t) == TOKEN_VARIABLE) {
		n1ind = t_variable(ct->t);
		ch = inch();
		tinybc_set(ct, n1ind, ch & A_CHARTEXT);
	}
	t_take(ct->t, TOKEN_VARIABLE);
	if (t_type(ct->t) == TOKEN_COMMA) {
		t_take(ct->t, TOKEN_COMMA);
		for (i = 1, k = 1; i <= 4; i++) {
			if (A_COLOR & k) break;
			if (i < 4) k *= 256;
		}
		co = (ch & A_COLOR) / k;
		for (i = 0; i < MAX_COLOR; i++) {
			for (j = 0; j < MAX_COLOR; j++)
				if (ct->colorpairs[i][j] == co)
					break;
			if (ct->colorpairs[i][j] == co) break;
		}
#ifdef BGR
		if (i == 1)
			i = 4;
		else if (i == 3)
			i = 6;
		else if (i == 4)
			i = 1;
		else if (i == 6)
			i = 3;
		if (j == 1)
			j = 4;
		else if (j == 3)
			j = 6;
		else if (j == 4)
			j = 1;
		else if (j == 6)
			j = 3;
#endif
		if (t_type(ct->t) == TOKEN_VARIABLE) {
			n2ind = t_variable(ct->t);
			tinybc_set(ct, n2ind, i);
		}
		t_take(ct->t, TOKEN_VARIABLE);
		t_take(ct->t, TOKEN_COMMA);
		if (t_type(ct->t) == TOKEN_VARIABLE) {
			n3ind = t_variable(ct->t);
			tinybc_set(ct, n3ind, j);
		}
		t_take(ct->t, TOKEN_VARIABLE);
	}
	t_take(ct->t, TOKEN_LF);
}

static void inkey_stmt(struct cttype *ct)
{
	long int n1ind;

	t_take(ct->t, TOKEN_INKEY);
	PDEBUG(lg, "inkey:\n");
	if (t_type(ct->t) == TOKEN_VARIABLE) {
		n1ind = t_variable(ct->t);
		tinybc_set(ct, n1ind, wgetch(stdscr));
	}
	t_take(ct->t, TOKEN_VARIABLE);
	refresh();
	t_take(ct->t, TOKEN_LF);
}

static void input_stmt(struct cttype *ct)
{
	char buffer[MAX_STRLEN], *p;
	long int n1ind, i, c, n;

	t_take(ct->t, TOKEN_INPUT);
	PDEBUG(lg, "input:\n");
	strcpy(buffer, "");
	printw("? ");
	interactive(0);
	refresh();
	getnstr(buffer, MAX_STRLEN - 1);
	p = buffer;
	for (; ;)
		if (t_type(ct->t) == TOKEN_VARIABLE) {
			n1ind = t_variable(ct->t);
			while (*p && !isdigit(*p) && *p != '-' &&
				(*p < 'A' || *p > 'Z'))
				p++;
			if (*p && (isdigit(*p) || *p == '-')) {
				tinybc_set(ct, n1ind, atoi(p));
				while (isdigit(*p) || *p == '-') p++;
			} else if (*p && *p >= 'A' && *p <= 'Z') {
				tinybc_set(ct, n1ind,
					tinybc_get(ct, *p - 'A'));
				while (isdigit(*p) || *p == '-') p++;
			}
			t_take(ct->t, TOKEN_VARIABLE);
		} else if (t_type(ct->t) == TOKEN_COMMA) {
			t_take(ct->t, TOKEN_COMMA);
		} else
			break;
	interactive(1);
	refresh();
	t_take(ct->t, TOKEN_LF);
}

static void let_stmt(struct cttype *ct)
{
	long int n1ind;

	n1ind = t_variable(ct->t);
	t_take(ct->t, TOKEN_VARIABLE);
	PDEBUG(lg, "let:\n");
	t_take(ct->t, TOKEN_EQ);
	tinybc_set(ct, n1ind, expr(ct));
	t_take(ct->t, TOKEN_LF);
}

static void locate_stmt(struct cttype *ct)
{
	long int n1, n2;

	t_take(ct->t, TOKEN_LOCATE);
	PDEBUG(lg, "locate:\n");
	if (t_expr_type(ct->t))
		n1 = expr(ct);
	else
		t_take(ct->t, TOKEN_VARIABLE);
	t_take(ct->t, TOKEN_COMMA);
	if (t_expr_type(ct->t)) {
		n2 = expr(ct);
		move(n1, n2);
	} else
		t_take(ct->t, TOKEN_VARIABLE);
	t_take(ct->t, TOKEN_LF);
}

static void nap_stmt(struct cttype *ct)
{
	long int n1;

	t_take(ct->t, TOKEN_NAP);
	PDEBUG(lg, "nap:\n");
	refresh();
	if (t_expr_type(ct->t)) {
		n1 = expr(ct);
		napms(n1);
	} else
		t_take(ct->t, TOKEN_VARIABLE);
	t_take(ct->t, TOKEN_LF);
}

static void out_stmt(struct cttype *ct)
{
	int n1;

	t_take(ct->t, TOKEN_OUT);
	PDEBUG(lg, "out:\n");
	n1 = (int) expr(ct);
	if (n1 != '\n')
		printw("%c", n1);
	else {
		scrollok(stdscr, TRUE);
		printw("\n");
		scrollok(stdscr, FALSE);
		refresh();
	}
	t_take(ct->t, TOKEN_LF);
}

static void print_stmt(struct cttype *ct)
{
	char string[MAX_STRLEN];
	int last;

	t_take(ct->t, TOKEN_PRINT);
	PDEBUG(lg, "print:\n");
	last = 0;
	for (; ;)
		if (t_type(ct->t) == TOKEN_STRING) {
			t_string(ct->t, string, MAX_STRLEN);
			printw("%s", string);
			last = TOKEN_STRING;
			t_take(ct->t, TOKEN_STRING);
		} else if (t_type(ct->t) == TOKEN_COMMA) {
			printw(" ");
			last = TOKEN_COMMA;
			t_take(ct->t, TOKEN_COMMA);
		} else if (t_type(ct->t) == TOKEN_SEMICOLON) {
			last = TOKEN_SEMICOLON;
			t_take(ct->t, TOKEN_SEMICOLON);
		} else if (t_expr_type(ct->t)) {
			last = TOKEN_NUMBER;
			printw("%d", expr(ct));
		} else
			break;
	if (last != TOKEN_COMMA && last != TOKEN_SEMICOLON) {
		scrollok(stdscr, TRUE);
		printw("\n");
		scrollok(stdscr, FALSE);
		refresh();
	}
	t_take(ct->t, TOKEN_LF);
}

static void rem_stmt(struct cttype *ct)
{
	t_take(ct->t, TOKEN_REM);
	PDEBUG(lg, "rem:\n");
	t_take(ct->t, TOKEN_LF);
}

static void return_stmt(struct cttype *ct)
{
	long int linenum;

	t_take(ct->t, TOKEN_RETURN);
	PDEBUG(lg, "return:\n");
	if (ct->gosub_stack_ptr > 0)
		linenum = ct->gosub_stack[--(ct->gosub_stack_ptr)];
	else {
		linenum = 0;
		printw("Gosub stack empty, line %d\n",
			t_currline(ct->t));
		t_end(ct->t);
		return;
	}
	t_jump(ct->t, linenum);
}

static void system_stmt(struct cttype *ct)
{
	char buffer[MAX_STRLEN], path[MAX_STRLEN], *p;
	long int n1, n2, pos, i;
	FILE *f;

	t_take(ct->t, TOKEN_SYSTEM);
	PDEBUG(lg, "assign:\n");
	if (t_expr_type(ct->t))
		n1 = expr(ct);
	else
		t_take(ct->t, TOKEN_VARIABLE);
	t_take(ct->t, TOKEN_COMMA);
	if (t_expr_type(ct->t)) {
		n2 = expr(ct);
		strcpy(path, "");
		p = getenv("TINYBC_TMP");
		if (p && strlen(p) < MAX_STRLEN) strcpy(path, p);
		sprintf(buffer, "%sTBC1", path);
		f = fopen(buffer, "w");
		for (pos = n2; tinybc_get(ct, MAX_VARIND + pos); ) {
			for (i = 0; tinybc_get(ct, MAX_VARIND + i) &&
				tinybc_get(ct, MAX_VARIND + i) != '\n' &&
				i < MAX_STRLEN - 2; i++)
				buffer[i] = tinybc_get(ct, MAX_VARIND +
					pos + i);
			if (buffer[i] == '\n') i++;
			buffer[i] = 0;
			fputs(buffer, f);
			pos += i;
		}
		fclose(f);
		for (i = 0; tinybc_get(ct, MAX_VARIND + n1 + i) &&
			i < MAX_STRLEN - 2 * strlen(path) - 16; i++)
			buffer[i] = tinybc_get(ct, MAX_VARIND + n1 + i);
		buffer[i] = 0;
		sprintf(buffer, "%s < %sTBC1 > %sTBC2", buffer, path, path);
		system(buffer);
		sprintf(buffer, "%sTBC2", path);
		f = fopen(buffer, "r");
		for (pos = n2; f && fgets(buffer, MAX_STRLEN, f); ) {
			for (i = 0; buffer[i]; i++)
				tinybc_set(ct, MAX_VARIND + pos + i,
					buffer[i]);
			tinybc_set(ct, MAX_VARIND + pos + i, 0);
			pos += i;
		}
		fclose(f);
		sprintf(buffer, "%sTBC1", path);
		remove(buffer);
		sprintf(buffer, "%sTBC2", path);
		remove(buffer);
	} else
		t_take(ct->t, TOKEN_VARIABLE);
	t_take(ct->t, TOKEN_LF);
}

/* Functions of the new statements must be added here */
static void stmt(struct cttype *ct)
{
	switch (t_type(ct->t)) {
	case TOKEN_AT:
		at_stmt(ct);
		break;
	case TOKEN_CLS:
		cls_stmt(ct);
		break;
	case TOKEN_COLOR:
		color_stmt(ct);
		break;
	case TOKEN_END:
		end_stmt(ct);
		break;
	case TOKEN_GOSUB:
		gosub_stmt(ct);
		break;
	case TOKEN_GOTO:
		goto_stmt(ct);
		break;
	case TOKEN_IF:
		if_stmt(ct);
		break;
	case TOKEN_INCHAR:
		inchar_stmt(ct);
		break;
	case TOKEN_INKEY:
		inkey_stmt(ct);
		break;
	case TOKEN_INPUT:
		input_stmt(ct);
		break;
	case TOKEN_VARIABLE:
		let_stmt(ct);
		break;
	case TOKEN_LOCATE:
		locate_stmt(ct);
		break;
	case TOKEN_NAP:
		nap_stmt(ct);
		break;
	case TOKEN_OUT:
		out_stmt(ct);
		break;
	case TOKEN_PRINT:
		print_stmt(ct);
		break;
	case TOKEN_REM:
		rem_stmt(ct);
		break;
	case TOKEN_RETURN:
		return_stmt(ct);
		break;
	case TOKEN_SYSTEM:
		system_stmt(ct);
		break;
	case TOKEN_LF:
		t_take(ct->t, TOKEN_LF);
		break;
	default:
		scrollok(stdscr, TRUE);
		printw("\n");
		scrollok(stdscr, FALSE);
		printw("Statement not implemented, line %d\n",
			t_currline(ct->t));
		t_end(ct->t);
	}
}

void tinybc_close(struct cttype *ct)
{
	free(ct->t);

	clear();
	refresh();
	delwin(stdscr);
	endwin();
}

int tinybc_ended(struct cttype *ct)
{
	if (t_ended(ct->t)) {
		PDEBUG(lg, "tinybc ended\n");

		refresh();
		while (wgetch(stdscr) < 0) napms(100);
		attrset(COLOR_PAIR(0));
		bkgd(COLOR_PAIR(0));
		clear();
		refresh();
		interactive(0);
	}
	return t_ended(ct->t);
}

long int tinybc_get(struct cttype *ct, int varind)
{
	long int *data;

	data = t_data(ct->t);
	if (varind >= 0 && (varind + 1) * 4 < t_size(ct->t))
		return data[varind];
	return 0;
}

void tinybc_init(struct cttype *ct, char *program, long int *numbers)
{
	struct ttype *t;
	void *p;

	/* Make sure that your calloc is thread-safe */
	t = (struct ttype *) calloc(sizeof(struct ttype), 1);
	ct->t = t;
	t_init(ct->t, program, numbers);
	srand(time(0));

	initscr();
	start_color();
	intrflush(stdscr, FALSE);
	interactive(0);
}

void tinybc_set(struct cttype *ct, long int varind, long int value)
{
	long int *data;

	data = t_data(ct->t);
	if (varind >= 0 && (varind + 1) * 4 < t_size(ct->t)) {
		data[varind] = value;
		PDEBUG(lg, "set: assign value %d to data element "
			"number %d\n", value, varind);
	}
}

void tinybc_start(struct cttype *ct)
{
	clear();
	refresh();

	interactive(1);
	t_start(ct->t);
	if (t_size(ct->t) < MAX_VARIND * 4) {
		printw("Not enough memory for data\n");
		t_end(ct->t);
	}
}

void tinybc_statement(struct cttype *ct)
{
	if (t_ended(ct->t)) return;
	PDEBUG(lg, "\n---------------- Line %d ----------------\n",
		t_number(ct->t));
	if (t_type(ct->t) != TOKEN_NUMBER) {
		t_end(ct->t);
		PDEBUG(lg, "Ended, line with no number\n");
		return;
	}
	t_currline_set(ct->t, t_number(ct->t));
	if (t_toolong(ct->t)) {
		printw("Ended, too long line, line %d\n",
			t_currline(ct->t));
		t_end(ct->t);
		return;
	}
	/* Line numbers are compulsory */
	t_take(ct->t, TOKEN_NUMBER);
	if (t_type(ct->t) == TOKEN_LET) t_take(ct->t, TOKEN_LET);
	stmt(ct);
	return;
}

